File: frontend/src/lib/grc.ts
// Client API du pôle GRC (/v1/grc) — registre de conformité déterministe :
// obligations, contrôles, constats + synthèse (plan de contrôle).
import { api } from "./api";

// --- Obligations -------------------------------------------------------------

export interface Obligation {
  id: string;
  tenant_id: string;
  reference: string;
  intitule: string;
  domaine: string;
  autorite: string;
  periodicite: string;
  echeance: string | null;
  base_legale: string;
  statut: string;
  country: string;
  created_at: string | null;
}

export interface ObligationInput {
  reference?: string;
  intitule: string;
  domaine?: string;
  autorite?: string;
  periodicite?: string;
  echeance?: string | null;
  base_legale?: string;
  statut?: string;
}

export async function listObligations(): Promise<Obligation[]> {
  return (await api<{ obligations: Obligation[] }>("/v1/grc/obligations")).obligations;
}

export async function createObligation(body: ObligationInput): Promise<Obligation> {
  return api<Obligation>("/v1/grc/obligations", { body });
}

export async function patchObligation(id: string, body: Partial<ObligationInput>): Promise<Obligation> {
  return api<Obligation>(`/v1/grc/obligations/${id}`, { method: "PATCH", body });
}

export async function deleteObligation(id: string): Promise<{ status: string }> {
  return api<{ status: string }>(`/v1/grc/obligations/${id}`, { method: "DELETE" });
}

// --- Contrôles -----------------------------------------------------------------

export interface Control {
  id: string;
  tenant_id: string;
  obligation_id: string | null;
  intitule: string;
  type_controle: string;
  frequence: string;
  responsable: string;
  derniere_execution: string | null;
  prochaine_execution: string | null;
  statut: string;
  country: string;
  created_at: string | null;
}

export interface ControlInput {
  obligation_id?: string | null;
  intitule: string;
  type_controle?: string;
  frequence?: string;
  responsable?: string;
  derniere_execution?: string | null;
  prochaine_execution?: string | null;
  statut?: string;
}

export async function listControls(obligationId?: string): Promise<Control[]> {
  const qs = obligationId ? `?obligation_id=${encodeURIComponent(obligationId)}` : "";
  return (await api<{ controls: Control[] }>(`/v1/grc/controls${qs}`)).controls;
}

export async function createControl(body: ControlInput): Promise<Control> {
  return api<Control>("/v1/grc/controls", { body });
}

export async function patchControl(id: string, body: Partial<ControlInput>): Promise<Control> {
  return api<Control>(`/v1/grc/controls/${id}`, { method: "PATCH", body });
}

export async function deleteControl(id: string): Promise<{ status: string }> {
  return api<{ status: string }>(`/v1/grc/controls/${id}`, { method: "DELETE" });
}

// --- Constats --------------------------------------------------------------

export interface Finding {
  id: string;
  tenant_id: string;
  obligation_id: string | null;
  control_id: string | null;
  intitule: string;
  gravite: string;
  statut: string;
  date_constat: string;
  echeance_correction: string | null;
  plan_action: string;
  responsable: string;
  country: string;
  created_at: string | null;
}

export interface FindingInput {
  obligation_id?: string | null;
  control_id?: string | null;
  intitule: string;
  gravite?: string;
  statut?: string;
  date_constat: string;
  echeance_correction?: string | null;
  plan_action?: string;
  responsable?: string;
}

export async function listFindings(obligationId?: string): Promise<Finding[]> {
  const qs = obligationId ? `?obligation_id=${encodeURIComponent(obligationId)}` : "";
  return (await api<{ findings: Finding[] }>(`/v1/grc/findings${qs}`)).findings;
}

export async function createFinding(body: FindingInput): Promise<Finding> {
  return api<Finding>("/v1/grc/findings", { body });
}

export async function patchFinding(
  id: string,
  body: Partial<Pick<FindingInput, "intitule" | "gravite" | "statut" | "echeance_correction" | "plan_action" | "responsable">>,
): Promise<Finding> {
  return api<Finding>(`/v1/grc/findings/${id}`, { method: "PATCH", body });
}

export async function deleteFinding(id: string): Promise<{ status: string }> {
  return api<{ status: string }>(`/v1/grc/findings/${id}`, { method: "DELETE" });
}

// --- Synthèse (plan de contrôle) --------------------------------------------

export interface EcheanceGrc {
  type: "obligation" | "controle";
  reference: string;
  libelle: string;
  date_limite: string;
  jours_restants: number;
}

export interface SyntheseConformite {
  nb_obligations: number;
  nb_obligations_actives: number;
  nb_obligations_sans_controle: number;
  taux_couverture: string;
  nb_controls: number;
  nb_controls_en_retard: number;
  nb_findings: number;
  nb_findings_ouverts: number;
  taux_conformite: string;
  findings_ouverts_par_gravite: { critique: number; majeur: number; mineur: number };
  obligations_par_domaine: Record<string, number>;
  echeances: EcheanceGrc[];
  alertes: string[];
}

export async function getPlanControle(horizonJours = 90): Promise<SyntheseConformite> {
  return api<SyntheseConformite>(`/v1/grc/plan-controle?horizon_jours=${horizonJours}`);
}

File: src/zolaos/api/v1/grc.py
"""GRC — Gouvernance / Risque / Conformité : registre de conformité (GRC-1).

Profil box. Système de référence léger : obligations réglementaires/contractuelles
→ contrôles internes → constats (findings). Endpoint `/plan-controle` : synthèse
déterministe de pilotage (couverture, retards, échéances, taux de conformité).
"""

from __future__ import annotations

from datetime import date
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from zolaos.agents.grc.conformite import (
    ControlLite,
    FindingLite,
    ObligationLite,
    synthese_conformite,
)
from zolaos.db.session import get_session
from zolaos.db.store_repo import (
    ControlRepository,
    FindingRepository,
    ObligationRepository,
)

router = APIRouter(prefix="/v1/grc", tags=["grc"])


# --------------------------------------------------------------------- schémas


class ObligationIn(BaseModel):
    reference: str = ""
    intitule: str
    domaine: str = "autre"
    autorite: str = ""
    periodicite: str = "ponctuelle"
    echeance: date | None = None
    base_legale: str = ""
    statut: str = "active"
    country: str = "cg"


class ObligationPatch(BaseModel):
    reference: str | None = None
    intitule: str | None = None
    domaine: str | None = None
    autorite: str | None = None
    periodicite: str | None = None
    echeance: date | None = None
    base_legale: str | None = None
    statut: str | None = None


class ControlIn(BaseModel):
    obligation_id: str | None = None
    intitule: str
    type_controle: str = "preventif"
    frequence: str = "ponctuel"
    responsable: str = ""
    derniere_execution: date | None = None
    prochaine_execution: date | None = None
    statut: str = "planifie"
    country: str = "cg"


class ControlPatch(BaseModel):
    obligation_id: str | None = None
    intitule: str | None = None
    type_controle: str | None = None
    frequence: str | None = None
    responsable: str | None = None
    derniere_execution: date | None = None
    prochaine_execution: date | None = None
    statut: str | None = None


class FindingIn(BaseModel):
    obligation_id: str | None = None
    control_id: str | None = None
    intitule: str
    gravite: str = "mineur"
    statut: str = "ouvert"
    date_constat: date
    echeance_correction: date | None = None
    plan_action: str = ""
    responsable: str = ""
    country: str = "cg"


class FindingPatch(BaseModel):
    intitule: str | None = None
    gravite: str | None = None
    statut: str | None = None
    echeance_correction: date | None = None
    plan_action: str | None = None
    responsable: str | None = None


# ------------------------------------------------------------------ obligations


@router.post("/obligations", status_code=status.HTTP_201_CREATED, summary="Créer une obligation")
async def create_obligation(
    body: ObligationIn, tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, Any]:
    rec = await ObligationRepository(session).create({**body.model_dump(), "tenant_id": tenant_id})
    await session.commit()
    return rec.to_dict()


@router.get("/obligations", summary="Lister les obligations")
async def list_obligations(
    tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, Any]:
    rows = await ObligationRepository(session).list(tenant_id=tenant_id)
    return {"obligations": [r.to_dict() for r in rows]}


@router.patch("/obligations/{obligation_id}", summary="Mettre à jour une obligation")
async def patch_obligation(
    obligation_id: str,
    body: ObligationPatch,
    tenant_id: str = "local",
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    rec = await ObligationRepository(session).update(
        obligation_id, tenant_id=tenant_id, fields=body.model_dump(exclude_none=True)
    )
    if rec is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="obligation_not_found")
    await session.commit()
    return rec.to_dict()


@router.delete("/obligations/{obligation_id}", summary="Supprimer une obligation")
async def delete_obligation(
    obligation_id: str, tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, str]:
    ok = await ObligationRepository(session).delete(obligation_id, tenant_id=tenant_id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="obligation_not_found")
    await session.commit()
    return {"status": "deleted"}


# --------------------------------------------------------------------- contrôles


@router.post("/controls", status_code=status.HTTP_201_CREATED, summary="Créer un contrôle")
async def create_control(
    body: ControlIn, tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, Any]:
    rec = await ControlRepository(session).create({**body.model_dump(), "tenant_id": tenant_id})
    await session.commit()
    return rec.to_dict()


@router.get("/controls", summary="Lister les contrôles")
async def list_controls(
    tenant_id: str = "local",
    obligation_id: str | None = None,
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    rows = await ControlRepository(session).list(tenant_id=tenant_id, obligation_id=obligation_id)
    return {"controls": [r.to_dict() for r in rows]}


@router.patch("/controls/{control_id}", summary="Mettre à jour un contrôle")
async def patch_control(
    control_id: str,
    body: ControlPatch,
    tenant_id: str = "local",
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    rec = await ControlRepository(session).update(
        control_id, tenant_id=tenant_id, fields=body.model_dump(exclude_none=True)
    )
    if rec is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="control_not_found")
    await session.commit()
    return rec.to_dict()


@router.delete("/controls/{control_id}", summary="Supprimer un contrôle")
async def delete_control(
    control_id: str, tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, str]:
    ok = await ControlRepository(session).delete(control_id, tenant_id=tenant_id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="control_not_found")
    await session.commit()
    return {"status": "deleted"}


# ----------------------------------------------------------------------- constats


@router.post("/findings", status_code=status.HTTP_201_CREATED, summary="Créer un constat")
async def create_finding(
    body: FindingIn, tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, Any]:
    rec = await FindingRepository(session).create({**body.model_dump(), "tenant_id": tenant_id})
    await session.commit()
    return rec.to_dict()


@router.get("/findings", summary="Lister les constats")
async def list_findings(
    tenant_id: str = "local",
    obligation_id: str | None = None,
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    rows = await FindingRepository(session).list(tenant_id=tenant_id, obligation_id=obligation_id)
    return {"findings": [r.to_dict() for r in rows]}


@router.patch("/findings/{finding_id}", summary="Mettre à jour un constat")
async def patch_finding(
    finding_id: str,
    body: FindingPatch,
    tenant_id: str = "local",
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    rec = await FindingRepository(session).update(
        finding_id, tenant_id=tenant_id, fields=body.model_dump(exclude_none=True)
    )
    if rec is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="finding_not_found")
    await session.commit()
    return rec.to_dict()


@router.delete("/findings/{finding_id}", summary="Supprimer un constat")
async def delete_finding(
    finding_id: str, tenant_id: str = "local", session: AsyncSession = Depends(get_session)
) -> dict[str, str]:
    ok = await FindingRepository(session).delete(finding_id, tenant_id=tenant_id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="finding_not_found")
    await session.commit()
    return {"status": "deleted"}


# ------------------------------------------------------------- plan de contrôle


@router.get("/plan-controle", summary="Synthèse de conformité (couverture, retards, échéances)")
async def plan_controle(
    tenant_id: str = "local",
    horizon_jours: int = 90,
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    obligations = await ObligationRepository(session).list(tenant_id=tenant_id)
    controls = await ControlRepository(session).list(tenant_id=tenant_id)
    findings = await FindingRepository(session).list(tenant_id=tenant_id)

    synth = synthese_conformite(
        [ObligationLite(**o.to_dict()) for o in obligations],
        [ControlLite(**c.to_dict()) for c in controls],
        [FindingLite(**f.to_dict()) for f in findings],
        today=date.today(),
        horizon_jours=horizon_jours,
    )
    return synth.model_dump(mode="json")
